src/lib.rs

```diff
@@ -1,117 +1,201 @@
-#![allow(non_upper_case_globals)]
-#![allow(non_camel_case_types)]
-#![allow(non_snake_case)]
+#![doc(html_root_url = "https://docs.rs/zcash_script/0.3.0")]
 #![allow(unsafe_code)]
-#![allow(unused_imports)]
-#![allow(clippy::unwrap_or_default)]
 
-// Use the generated C++ bindings
-include!(concat!(env!("OUT_DIR"), "/bindings.rs"));
+mod cxx;
+pub use cxx::*;
+
+mod interpreter;
+pub use interpreter::{HashType, VerificationFlags};
+mod zcash_script;
+pub use zcash_script::*;
+
+use std::os::raw::{c_int, c_uint, c_void};
+
+/// A tag to indicate that the C++ implementation of zcash_script should be used.
+pub enum Cxx {}
+
+impl From<zcash_script_error_t> for Error {
+    #[allow(non_upper_case_globals)]
+    fn from(err_code: zcash_script_error_t) -> Error {
+        match err_code {
+            zcash_script_error_t_zcash_script_ERR_OK => Error::Ok,
+            zcash_script_error_t_zcash_script_ERR_VERIFY_SCRIPT => Error::VerifyScript,
+            unknown => Error::Unknown(unknown),
+        }
+    }
+}
+
+/// The sighash callback to use with zcash_script.
+extern "C" fn sighash_callback(
+    sighash_out: *mut u8,
+    sighash_out_len: c_uint,
+    ctx: *const c_void,
+    script_code: *const u8,
+    script_code_len: c_uint,
+    hash_type: c_int,
+) {
+    let checked_script_code_len = usize::try_from(script_code_len)
+        .expect("This was converted from a `usize` in the first place");
+    // SAFETY: `script_code` is created from a Rust slice in `verify_callback`, passed through the
+    // C++ code, eventually to `CallbackTransactionSignatureChecker::CheckSig`, which calls this
+    // function.
+    let script_code_vec =
+        unsafe { std::slice::from_raw_parts(script_code, checked_script_code_len) };
+    let ctx = ctx as *const SighashCalculator;
+    // SAFETY: `ctx` is a valid `SighashCalculator` passed to `verify_callback` which forwards it to
+    // the `CallbackTransactionSignatureChecker`.
+    if let Some(sighash) = unsafe { *ctx }(script_code_vec, HashType::from_bits_retain(hash_type)) {
+        assert_eq!(sighash_out_len, sighash.len().try_into().unwrap());
+        // SAFETY: `sighash_out` is a valid buffer created in
+        // `CallbackTransactionSignatureChecker::CheckSig`.
+        unsafe { std::ptr::copy_nonoverlapping(sighash.as_ptr(), sighash_out, sighash.len()) };
+    }
+}
+
+/// This steals a bit of the wrapper code from zebra_script, to provide the API that they want.
+impl ZcashScript for Cxx {
+    fn verify_callback(
+        sighash: SighashCalculator,
+        lock_time: i64,
+        is_final: bool,
+        script_pub_key: &[u8],
+        signature_script: &[u8],
+        flags: VerificationFlags,
+    ) -> Result<(), Error> {
+        let mut err = 0;
+
+        // SAFETY: The `script` fields are created from a valid Rust `slice`.
+        let ret = unsafe {
+            zcash_script_verify_callback(
+                (&sighash as *const SighashCalculator) as *const c_void,
```

Comment:
I still feel like these casts are the wrong way to get this pointer.

Reply:
The problem is that there’s no way to do it otherwise (AFAIK) that allows me to avoid changing the C++ code. I know it’s not sacred, but I’d like to at least get in something that works with no C++ changes first.

If I `Box`, then pointers through a bunch of C++ files lose their `const`. To me, preserving `const` seems better than switching to `Box`, but I’m certainly no expert here.

Reply:
> I still feel like these casts are the wrong way to get this pointer.

It doesn't look great, but it's the same as it was in `zebra_script` (except without dereferencing the `Box`, which seems like an improvement), it's just been moved here from `zebra_script`, so if it's possible to change it, I think we should do that in another PR.

Reply:
I just realized what bothers me about it. This is a pointer to the callback's memory on the _stack_. That makes me uncomfortable; whenever we cross the FFI, we make sure to use heap memory.

Reply:
> whenever we cross the FFI, we make sure to use heap memory

What's the advantage of using heap memory whenever we cross the FFI?

Reply:
Giving a pointer to stack memory to a foreign process then gives that process the opportunity to choose an offset from that pointer that gives them control up the stack, which is generally a bad idea.

Reply:
This is a function pointer though, I don't think it points to the stack. I couldn't find much information about how Rust handles Rust function pointers across FFI and back though. (For reference my original code was based on https://doc.rust-lang.org/nomicon/ffi.html#targeting-callbacks-to-rust-objects)

I'm not sure I get your point. It's the same process, not a foreign process. The C code could walk up the stack regardless.

Reply:
Ok, I’m sticking with the `const` for now. I’m not a fan of the raw pointers, but
1. removing `const` requires changing the C++ code, which I specifically wanted to avoid (at least initially) to ensure that there is no chance of behavioral change there),
2. `const` is Good, and
3. it’s what the original code in Zebra was doing, so again, just one less thing that differs in this PR.

@conradoplg Could you approve the workflows? Or are you not technically a maintainer, and I need someone else? (@upbqdn?)

Reply:
Oops, forgot about it, just approve it.
I'm happy with merging the current approach but might be wise at some point to consult with someone more familiar with FFI to double check.